### PR TITLE
Fix poll ready stop delays

### LINKED ISSUE #166: pgctl-poll-ready causes pgctl stop to timeout
`pgctl-poll-ready` is used to implement most healthchecks for services in pgctl when those services aren't quick to startup. It is designed to calls the ready script repeatedly until the service reports ready, and then report that the service is up over a file descriptor. Unfortunately, there is a bug in its busy waiting logic, which causes pgctl to think a process has been left behind.

`pgctl-poll-ready` forks a background process that actually implements the behavior of the tool, and then execs the actual service. In the background, `pgctl-poll-ready` has to do two things once a service is up: a) periodically run ready to make sure the service is still alive (once every 10 seconds), and b) if `pgctl stop` is run, end the background process. Unfortunately, (b) is slow because the background process is single threaded and in that thread, a sleep for (a) is usually where execution is waiting. This means that even if `pgctl stop` is invoked, it will have to finish waiting for that sleep to complete before it wakes up and realizes that it should exit. Because it takes a long time for the background process to die, pgctl to become unhappy and throw errors about processes being left behind by a stopped service.

@bukzor found that we could fix this issue by using a timeout on the wait for condition (b) as our sleep for condition (a). We can use the -t option to [s6-ftrig-wait](https://skarnet.org/software/s6/s6-ftrig-wait.html) to specify that timeout.

### PR DESCRIPTION
**Pull request by 0xngold**
This fixes #166.

### CODE REVIEW

File: pgctl/poll_ready.py [hunk 1]
```diff
@@ -10,12 +10,16 @@
...
 import os
 import os.path
+import select
 import time
 
 from .functions import exec_
 from .functions import print_stderr
 
 
+DOWN_FIFO_PATH = os.path.join('event', 'ftrig1' + 'poll_ready'.ljust(43, '_'))
```
Comment (bukzor) [inline pgctl/poll_ready.py]: Should add a comment, referencing https://github.com/skarnet/s6/blob/v2.2.2.0/src/libs6/ftrigw_notifyb_nosig.c#L29,L30

File: pgctl/poll_ready.py [hunk 2]
```diff
@@ -88,14 +113,22 @@ def main():
...
-        down_event = Popen(
-            ('s6-ftrig-wait', 'event', 'd'),
-            stdout=open(os.devnull, 'w'),  # this prints 'd' otherwise
-        )
-
-        return pgctl_poll_ready(down_event, notification_fd, timeout, poll_ready, poll_down)
+
+        # Don't reuse an old FIFO
+        if os.path.exists(DOWN_FIFO_PATH):
+            os.remove(DOWN_FIFO_PATH)
```
Comment (bukzor) [inline pgctl/poll_ready.py]: This is racy. Prefer `try:remove`.

or even Popen(rm -f)

File: pgctl/poll_ready.py [hunk 3]
```diff
@@ -88,14 +113,22 @@ def main():
...
+
+        # FIFO listens for the s6 down event
+        #
+        # Even though the FIFO is effectively RO, it is opened as RW because
+        # opening as RO blocks until the other side of the FIFO is opened.
+        os.mkfifo(DOWN_FIFO_PATH)
```
Comment (bukzor) [inline pgctl/poll_ready.py]: Is there any case where the event/ directory doesn't exist?

File: pgctl/poll_ready.py [hunk 4]
```diff
@@ -88,14 +113,22 @@ def main():
...
+        down_fifo = os.open(DOWN_FIFO_PATH, os.O_RDWR)
+
+        pgctl_poll_ready(down_fifo, notification_fd, timeout, poll_ready, poll_down)
+
+        os.close(down_fifo)
+        os.remove(DOWN_FIFO_PATH)
```
Comment (bukzor) [inline pgctl/poll_ready.py]: You want:

```
try:
    return pgctl_poll_ready()
finally:
    os.close
    os.remove
```

File: tests/spec/poll_ready_quick_shutdown.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+def it_stops_quickly():
+    """Tests a regression in pgctl where services using pgctl-poll-ready fail to
+    stop because the background process started by pgctl-poll-ready isn't dying
+    quickly."""
+    check_call(('pgctl', 'start'))
+    prestop_time = time.time()
+    check_call(('pgctl', 'stop'))
+    poststop_time = time.time()
+    assert poststop_time - prestop_time < 2
```
Comment (bukzor) [inline tests/spec/poll_ready_quick_shutdown.py]: This fails without the patch yea?

Comment (0xngold) [inline tests/spec/poll_ready_quick_shutdown.py]: Yeah it does.

File: pgctl/poll_ready.py (hunk 1 above)
Comment (0xngold) [inline pgctl/poll_ready.py]: Agreed, fixing it.

File: pgctl/poll_ready.py (hunk 2 above)
Comment (0xngold) [inline pgctl/poll_ready.py]: Good catch. Fixing it.

File: pgctl/poll_ready.py (hunk 3 above)
Comment (0xngold) [inline pgctl/poll_ready.py]: I believe that isn't possible. The control flow should be pgctl -> S6 -> poll ready, and S6 creates that fifodir.

File: pgctl/poll_ready.py (hunk 4 above)
Comment (0xngold) [inline pgctl/poll_ready.py]: I do indeed.

**Pull request merged**